index.js

```diff
@@ -25,102 +25,111 @@ class Turtler {
+    if(!this.asciiTable) {
+      const { data, hasHeader, columnSeparator, headerSeparator } = this;
+
+      let table = '';
+      let columns = 0;
+      let columnWidths = [];
+
+      // Find the maximum width of each column
+      // If rows contain uneven number of columns, throw
+      data.forEach((row) => {
+        // The row should be an array
+        if(!Array.isArray(row)) throw new Error('data should be an array of arrays');
+        // Set the initial length of the row
+        if(!columns) columns = row.length;
+        // If the current row is not the same length as the initial one throw error
+        if(columns !== row.length) throw new Error('columns are not formed properly');
+
+        // find the maximum length of each column
+        row.forEach((v, l) => {
+          // column values must be strings
+          if(typeof v !== 'string') throw new Error('column values should be strings');
+
+          // Find the maximum string length in each column
+          if(!columnWidths[l] || columnWidths[l] < v.length) {
+            columnWidths[l] = v.length;
+          }
+        });
       });
-    });
-
-    data.forEach((row, l) => {
-      row = row.map((value, i) => {
-        // Create pad of empty spaces to match the width of this value to max width of this column
-        let padding = ' '.repeat(columnWidths[i] - value.length);
-        return value + padding;
-      // join on the columnSeparator
-      }).join(columnSeparator);
 
-      table += `${row}\n`;
+      data.forEach((row, l) => {
+        row = row.map((value, i) => {
+          // Create pad of empty spaces to match the width of this value to max width of this column
+          let padding = ' '.repeat(columnWidths[i] - value.length);
+          return value + padding;
+        // join on the columnSeparator
+        }).join(columnSeparator);
 
-      if(l === 0) {
-        // ignore the header string if hasHeader is false or headerSeparator is not set
-        if(!hasHeader || !headerSeparator) return;
+        table += `${row}\n`;
 
-        // we add columnSeparator.length because above we joined on that string which could be more than one character.
-        // We only want one character so we only look at the first character of the string
-        table += headerSeparator[0].repeat((columnWidths.reduce((a, b) => a + b)) + columnSeparator.length) + '\n';
-      }
-    });
+        if(l === 0) {
+          // ignore the header string if hasHeader is false or headerSeparator is not set
+          if(!hasHeader || !headerSeparator) return;
 
-    return table;
+          // we add columnSeparator.length because above we joined on that string which could be more than one character.
+          // We only want one character so we only look at the first character of the string
+          table += headerSeparator[0].repeat((columnWidths.reduce((a, b) => a + b)) + columnSeparator.length) + '\n';
+        }
+      });
+      this.asciiTable = table;
```

Comment:
good idea on the cache, problem will be if the data changes, the table should also change. 

I could override the table data using

```js
const Turtler = require('turtler');

let table = new Turtler([
  ["uid", "name"],
  ["1", "Doe"],
  ["2", "Hemma"]
]);

table.ascii()

table.data = [
  ["uid", "name"],
  ["1", "New Name"],
  ["2", "Hemma"],
  ["3", "new Row"]
]

table.ascii();
```

Reply:
we may want to memoize these functions:

https://medium.freecodecamp.org/understanding-memoize-in-javascript-51d07d19430e

Reply:
No No No. Data shouldn't change after an object is created. You want change in data, create another object.

Reply:
@anusaini but we can have a mutable function something like:

```js
const Turtler = require('turtler');

let table = new Turtler([
  ["uid", "name"],
  ["1", "Doe"],
  ["2", "Hemma"]
]);

table.ascii()

table.set([
  ["uid", "name"],
  ["1", "New Name"],
  ["2", "Hemma"],
  ["3", "new Row"]
]);
```

I think this should be allowed. Any reason to not be able to mutate data?

Reply:
Depends upon style and goals really. I prefer for objects to not change once created; that provides consistency and way to grep/reason about.

Any latest change to table data would actually be `new Table(...)`.

Reply:
@anusaini I agree, most times I usually only create one table object and then never go back and try to call toString on it. The best way of caching without having any idea of what the developer will do is to use `memoization`. We simply wrap the call in a cache method and bam, never have to worry about it.

Reply:
`Memoization` is helpful when you are going to depend upon earlier cached results (; say dynamic programming).

In this case, do you think user/developer will actually be switching data back-and-forth frequently.


```
let table = new Table (inputdata_a);
// operate on table
table.set(inputdata_b);
// operate on table
table.set(inputdata_a);
// operate on table
table.set(inputdata_b);
// operate on table
table.set(inputdata_a);
// operate on table
// so on
```

OR do you want to memoize on individual rows? (shudder! and excited?) ;)

Reply:
ohhh good point, @anusaini it really brings into question how useful a cache will be, maybe if the table data changes we just blowup the cache?

Reply:
Yes. And object will hold data/memory that cannot be GCed even though app might not use it ever.

On the other hand, calling `toString()`, `markdown()`, and `ascii()` to generate table every time looks like overkill. Think 1 MiB data getting parsed and generated every time you need it.

Reply:
I agree, so what do you think the best way to cache would be? If the user knows they need to access the data, can't they be the one to store it? They would still need to store the Turtler instance in any case.

Reply:
Immutable object with output caching to save from future calculations. That's where I am at. Setters are icky. 

You make a decision now, so that we can move forward in whatever direction.